server/docker_resource.py

```diff
@@ -0,0 +1,433 @@
+            if isinstance(val, dict):
+                # assume receiving data in the format of DockerCache
+                # validate all data
+                newSettings = DockerCache(val)
+                DockerResource.validateDict(newSettings)
+            elif isinstance(val, list):
+                # assume this is a list of desired images
+
+                for img in val:
+
+                    if not isinstance(img, six.string_types):
+                        raise RestException('%s was not a valid string.' % img)
+                else:
+                    # do not want to save the list as the setting
+
+                    DockerResource.appendImageJob(val)
+
+            elif isinstance(val, six.string_types):
+                # do not want to save the string as the setting
+
+                DockerResource.appendImageJob([val])
+
+            else:
+
+                raise RestException('name was not a valid JSON list or string.')
+
+            event.info['value'] = currentSettings.raw()
+            event.preventDefault().stopPropagation()
+
+    @staticmethod
+    def validateDict(newSettings):
+        currentSettings = DockerCache(getDockerImageSettings())
+        job = ModelImporter.model('job', 'jobs').createLocalJob(
+            module='girder.plugins.HistomicsTK.image_worker',
+            function='verifyDictionary',
+            kwargs={
+                'newSettings': newSettings.raw(),
+                'oldSettings': currentSettings.raw()
+            },
+
+            title='Updating Settings ',
+            type='HistomicsTK.images',
+            user=getCurrentUser(),
+            public=True,
+            async=True
+        )
+        ModelImporter.model('job', 'jobs').scheduleJob(job)
+
+
+def getDockerImageSettings():
+    module_list = ModelImporter.model('setting').get(
+        PluginSettings.DOCKER_IMAGES)
+    if module_list is None:
+        module_list = {}
+    return module_list
+
+
+class DockerCache():
+
+    imageName = 'docker_image_name'
+    type = 'type'
+    xml = 'xml'
+    cli_dict = 'cli_list'
+
+    def __init__(self, cache):
+        """
+        Data is stored in the following format:
+            {image_name_hash:
+                {
+                cli_list:{
+                    cli_name:{
+                        type:<type>
+                        xml:<xml>
+
+                            }
+                        }
+                docker_image_name:<name>
+                }
+            }
+        """
+        if isinstance(cache, dict):
+            self.data = cache
+        elif isinstance(cache, str):
+            if cache != '':
+                self.data = json.loads(cache)
+        else:
+            self.data = {}
+
+    def getDockerImageList(self):
+
+        return [str(imgDict[DockerCache.imageName])
+                for (imgHash, imgDict) in iteritems(self.data)]
+
+    def imageAlreadyLoaded(self, name):
```

Comment:
We might also find a method called `hasCLI` useful. You can add it in the next PR in the DockerImageCache class that manages a list of CLIs